### PR TITLE
Update and fix MS.DEFENDER.4.3-4.6 for newest baseline

### LINKED ISSUE #123: Defender 2.2 Policy should be set to "block everyone"
# 🐛 Summary #

Per the baseline, "The action for the DLP policy SHOULD be set to block sharing sensitive information with everyone when DLP conditions are met."

That means, the action for a given DLP rule should look like this:
image

However, the code currently also accepts "Block only people outside your organization."

## To reproduce ##

Follow the instructions [here](https://github.com/cisagov/ScubaGear/blob/main/baselines/defender.md#224-implementation), but select "Block only people outside your organization", then run ScubaGear. Note that the test still passes.

## Expected behavior ##
The relevant variables the Rego should check are `BlockAccess` and `BlockAccessScope`. The code already asserts that `BlockAccess` is set to true, but `BlockAccessScope` needs to be checked as well. It will be set to `All` if "Block everyone" was selected and "PerUser" if "Block only people outside your organization" was selected.

### PR DESCRIPTION
**Pull request by schrolla**
#  #

## 🗣 Description ##

The emerald version of the M365 Defender baseline includes significant updates to policies for data loss prevention.  As a result, the assessment code also needs to be updated.  

Major updates in this request include:
* Add SensitiveInfoPolicies test for 4.1 matching DLP policies
* Ensure 4.2/4.3/4.4 fail if 4.1 fails via custom error messages
* Fix bug where 4.3 would pass when action is `Block People Outside of Organization` rather than `Block Everyone`
* Update unit tests to note failure when 4.1 DLP policies not present
* Convert and add 4.3 and 4.4 unit tests
* Convert existing 4.5 and 4.6 not implemented unit tests to new format

## 💭 Motivation and context ##

All tests in the group are dependent on 4.1 passing such that a custom policy exists that contains at least the minimum sensitive info types (SSN, ITIN, and CC).  If no such custom policy exists, then all following policies in the group should also fail as only policies matching 4.1 are relevant for 4.2-4.6.

Closes #123 

## 🧪 Testing ##

Code was tested through passing all unit tests via `Testing\RunUnitTests.ps1 -p defender -c 4`

Limited functional testing was done via the G3G5/G5(high) tenants to ensure results as intended.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Rebase branch from any changes in parent branch and ensure tests still pass

## ✅ Post-merge checklist ##

- [ ] Check parent branch to ensure tests are still passing

### CODE REVIEW

**Review by nanda-katikaneni — COMMENTED** [review]
Doing functional tests for policy 4.x with different config combinations on G5 and E5 tenants - will update after the tests.

<redacted earlier comment - was looking into wrong branch>.

**Review by nanda-katikaneni — APPROVED** [review]
Reviewed the changes and tested it on G5 and E5 tenants (different config combinations for compliant and non-complaint cases) - got correct results for all cases. Approving.

File: Rego/DefenderConfig.rego [hunk 1]
```diff
@@ -421,6 +421,62 @@ tests[{
...
+    Policy := input.dlp_compliance_policies[_]
+
+    # Inspect rules for sensitive info types
+    InfoTypes := SensitiveInfoTypes(Policy.Name)
+
+    Conditions := [ "U.S. Social Security Number (SSN)" in InfoTypes,
+                    "U.S. Individual Taxpayer Identification Number (ITIN)" in InfoTypes,
+                    "Credit Card Number" in InfoTypes]
+
+    count([Condition | Condition = Conditions[_]; Condition == true]) == 3
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:475]: unification -> assignment
```suggestion
    count([Condition | Condition := Conditions[_]; Condition == true]) == 3
```

File: Rego/DefenderConfig.rego [hunk 2]
```diff
@@ -605,9 +667,11 @@ tests[{
...
     "RequirementMet": Status
 }] {
-    error_policy := "No enabled policy found that applies to:"
-    ErrorMessage := concat(" ", [error_policy, concat(", ", error_policies)])
-    Status := count(error_policies) == 0
+    ErrorMessage := DefenderErrorMessage42
+    Conditions := [count(error_policies) == 0,
+                   count(SensitiveInfoPolicies) > 0]
+    Status := count([Condition | Condition := Conditions[_];
+                     Condition == true ]) == 2
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:674]: Why are these on separate lines?
```suggestion
    Status := count([Condition | Condition := Conditions[_]; Condition == true ]) == 2
```

File: Rego/DefenderConfig.rego [hunk 3]
```diff
@@ -605,9 +667,11 @@ tests[{
     "ReportDetails": CustomizeError(ReportDetailsBoolean(Status), ErrorMessage),
     "RequirementMet": Status
 }] {
-    error_policy := "No enabled policy found that applies to:"
-    ErrorMessage := concat(" ", [error_policy, concat(", ", error_policies)])
-    Status := count(error_policies) == 0
+    ErrorMessage := DefenderErrorMessage42
+    Conditions := [count(error_policies) == 0,
+                   count(SensitiveInfoPolicies) > 0]
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:672]: Clean up spacing
```suggestion
    Conditions := [ count(error_policies) == 0,
                    count(SensitiveInfoPolicies) > 0 ]
```

File: Rego/DefenderConfig.rego [hunk 4]
```diff
@@ -617,42 +681,39 @@ tests[{
...
 }
 
-SensitiveRulesNotBlocking[Rule.Name] {
-    Rule := SensitiveRules[_]
-    Rule.BlockAccess                              
-    Policy := input.dlp_compliance_policies[_]
-    Rule.ParentPolicyName == Policy.Name
-    count(error_rules) == 0
-    Rule.BlockAccessScope != "All"
+DefenderErrorMessage43(Rules) := GenerateArrayString(Rules, ErrorMessage) if {
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:694]: Nitpick: This is not the 43rd Defender Error message, this is the Error Message for Defender 4.3; Therefore, `DefenderErrorMessage4_3`

File: Rego/DefenderConfig.rego [hunk 5]
```diff
@@ -617,42 +681,39 @@ tests[{
...
+    "ReportDetails" : CustomizeError(ReportDetailsBoolean(Status), ErrorMessage),
     "RequirementMet" : Status
 }] {
     Rules := SensitiveRulesNotBlocking
-    ErrorMessage := "rule(s) found that do(es) not block access or associated policy not set to enforce block action:"
-    Status := count(Rules) == 0
+    ErrorMessage := DefenderErrorMessage43(Rules)
+    Conditions := [count(Rules) == 0, count(SensitiveInfoPolicies) > 0]
+    Status := count([Condition | Condition := Conditions[_];
+                                    Condition == true ]) == 2
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:716]: Weird spacing again
```suggestion
    Conditions := [ count(Rules) == 0, 
                    count(SensitiveInfoPolicies) > 0 ]
    Status := count([Condition | Condition := Conditions[_]; Condition == true ]) == 2
```

File: Rego/DefenderConfig.rego [hunk 6]
```diff
@@ -662,20 +723,34 @@ tests[{
 # Step 4: ensure that some user is notified in the event of a DLP violation
 SensitiveRulesNotNotifying[Rule.Name] {
     Rule := SensitiveRules[_]
+    count(SensitiveInfoPolicies) > 0
+    Rule.ParentPolicyName in SensitiveInfoPolicies
     count(Rule.NotifyUser) == 0
 }
 
+DefenderErrorMessage44(Rules) := GenerateArrayString(Rules, ErrorMessage) if {
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:731]: Same Nitpick: This is not the 44th Defender Error message, this is the Error Message for Defender 4.4; Therefore, `DefenderErrorMessage4_4`

File: Rego/DefenderConfig.rego [hunk 7]
```diff
@@ -662,20 +723,34 @@ tests[{
...
+    "ReportDetails" : CustomizeError(ReportDetailsBoolean(Status), ErrorMessage),
     "RequirementMet" : Status
 }] {
     Rules := SensitiveRulesNotNotifying
-    ErrorMessage := "rule(s) found that do(es) not notify at least one user:"
-    Status := count(Rules) == 0
+    ErrorMessage := DefenderErrorMessage44(Rules)
+    Conditions := [count(Rules) == 0, count(SensitiveInfoPolicies) > 0]
+    Status := count([Condition | Condition := Conditions[_];
+                                    Condition == true ]) == 2
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:753]: Weird spacing again
```suggestion
    Conditions := [ count(Rules) == 0, 
                    count(SensitiveInfoPolicies) > 0 ]
    Status := count([Condition | Condition := Conditions[_]; Condition == true ]) == 2
```

File: Rego/DefenderConfig.rego [hunk 8]
```diff
@@ -421,6 +421,62 @@ tests[{
...
+
+    # Inspect rules for sensitive info types
+    InfoTypes := SensitiveInfoTypes(Policy.Name)
+
+    Conditions := [ "U.S. Social Security Number (SSN)" in InfoTypes,
+                    "U.S. Individual Taxpayer Identification Number (ITIN)" in InfoTypes,
+                    "Credit Card Number" in InfoTypes]
+
+    count([Condition | Condition = Conditions[_]; Condition == true]) == 3
+    Policy.Enabled = true
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:476]: This is the unification operator, not checking. If it is supposed to check `==`, if it is supposed to set `:=`.
```suggestion
    Policy.Enabled == true
```

File: Rego/DefenderConfig.rego [hunk 9]
```diff
@@ -421,6 +421,62 @@ tests[{
...
+    Policy.Mode == "Enable"
```
Comment (Sloane4) [inline Rego/DefenderConfig.rego:477]: SensitiveInfoPolicies is used in all Policies part of the 4 group. But I did not see unit test for line 476 in 4.1 & 4.3. I did not see unit test for failing line 477 in 4.1 & 4.2.

**Review by Sloane4 — APPROVED** [review]

**Pull request merged**